### PR TITLE
Add NodePostprocessor to NetworkRetriever & NetworkRetriever Demo

### PR DESCRIPTION
**Pull request by nerdai**
# Description

- This PR adds an optional NodePostprocessor to the NetworkRetriever
- Also adds a new demo for using NetworkRetriever
     - Privacy-Safe network exchange using differentially-private synthetic data 

Fixes # (issue)

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (TODO)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama-index-networks/pyproject.toml [hunk 1]
```diff
@@ -32,7 +32,7 @@ maintainers = [
 name = "llama-index-networks"
 packages = [{include = "llama_index/"}]
 readme = "README.md"
-version = "0.2.0"
+version = "0.2.1a2"
```
Comment (logan-markewich) [inline llama-index-networks/pyproject.toml:35]: Do you want this to stay as an alpha?

File: llama-index-networks/llama_index/networks/network/retriever.py [hunk 2]
```diff
@@ -15,17 +16,36 @@ class NetworkRetriever(BaseRetriever):
     def __init__(
         self,
         contributors: List[ContributorRetrieverClient],
+        reranker: Optional[BaseNodePostprocessor] = None,
```
Comment (logan-markewich) [inline llama-index-networks/llama_index/networks/network/retriever.py:19]: Couldn't this be any node post processor? Not just a reranker?

File: llama-index-networks/llama_index/networks/network/retriever.py [hunk 3]
```diff
@@ -15,17 +16,36 @@ class NetworkRetriever(BaseRetriever):
...
+        rerank: bool = False,
```
Comment (logan-markewich) [inline llama-index-networks/llama_index/networks/network/retriever.py:20]: Just curious, why we need both to provide a reranker and enable reranking?

**Review by logan-markewich — APPROVED** [review]

File: llama-index-networks/llama_index/networks/network/retriever.py (hunk 3 above)
Comment (nerdai) [reply llama-index-networks/llama_index/networks/network/retriever.py:20]: yeah, don't need both! will just use `reranker` if exists, then rerank :)

File: llama-index-networks/llama_index/networks/network/retriever.py (hunk 2 above)
Comment (nerdai) [reply llama-index-networks/llama_index/networks/network/retriever.py:19]: I guess it could be yeah. I'll just rename this to `node_postprocessor`

Comment (nerdai) [reply llama-index-networks/llama_index/networks/network/retriever.py:19]: aha, found [this](https://github.com/run-llama/llama_index/blob/ef54ac6b7dcf9b91b2278a66d5699f1aa1eb70c1/llama-index-core/llama_index/core/chat_engine/condense_plus_context.py#L169) in our code base - will adopt this pattern as well.

File: llama-index-networks/pyproject.toml (hunk 1 above)
Comment (nerdai) [reply llama-index-networks/pyproject.toml:35]: thanks! nope this needs to be upgraded to `0.2.1`

**Comment by nerdai** [comment]
Really have no idea why `test_openai_agent.py` is failing here...

File: llama-index-networks/llama_index/networks/network/retriever.py [hunk 4]
```diff
@@ -34,4 +47,14 @@ async def _aretrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
         contributor_results: List[List[NodeWithScore]] = await asyncio.gather(
             *async_tasks
         )
-        return reduce(lambda x, y: x + y, contributor_results)
+        nodes = reduce(lambda x, y: x + y, contributor_results)
+
+        for postprocessor in self._node_postprocessors:
+            nodes = postprocessor.postprocess_nodes(nodes, query_bundle=query_bundle)
```
Comment (logan-markewich) [inline llama-index-networks/llama_index/networks/network/retriever.py:53]: Note to self: add async node post processors lol

**Review by logan-markewich — APPROVED** [review]

File: llama-index-networks/llama_index/networks/network/retriever.py (hunk 4 above)
Comment (nerdai) [reply llama-index-networks/llama_index/networks/network/retriever.py:53]: Haha. Of course, how did I miss that. I'll add it quickly.

Comment (nerdai) [reply llama-index-networks/llama_index/networks/network/retriever.py:53]: oh nvm we don't have this yet?

**Pull request merged**